Objects/exceptions.c

```diff
@@ -629,6 +629,454 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
+    if (PySequence_Check(excs)) {
+        numexcs = PySequence_Length(excs);
+    }
+    if (numexcs <= 0) {
+        PyErr_SetString(PyExc_TypeError, "Expected a non-empty sequence");
+        return NULL;
+    }
+
+    int nested_base_exceptions = 0;
+    for (Py_ssize_t i = 0; i < numexcs; i++) {
+        PyObject *exc = PySequence_GetItem(excs, i);
+        if (!exc) {
+            return NULL;
+        }
+        if (!PyExceptionInstance_Check(exc)) {
+            PyErr_SetString(PyExc_TypeError,
+                "Nested exceptions must derive from BaseException");
+            Py_DECREF(exc);
+            return NULL;
+        }
+        int is_exception = PyObject_IsInstance(exc, PyExc_Exception);
+        Py_DECREF(exc);
+        if (is_exception < 0) {
+            return NULL;
+        }
+        else if (is_exception == 0) {
+            nested_base_exceptions = 1;
+        }
+    }
+
+    PyTypeObject *cls = type;
+    if (cls == (PyTypeObject*)PyExc_ExceptionGroup) {
+        if (nested_base_exceptions) {
+            PyErr_SetString(PyExc_TypeError,
+                "Cannot nest BaseExceptions in an ExceptionGroup");
+            return NULL;
+        }
+    }
+    else if (cls == (PyTypeObject*)PyExc_BaseExceptionGroup) {
+        if (!nested_base_exceptions) {
+            /* All nested exceptions are Exception subclasses,
+             * wrap them in an ExceptionGroup
+             */
+            cls = (PyTypeObject*)PyExc_ExceptionGroup;
+        }
+    }
+    else {
+        /* Do nothing - we don't interfere with subclasses */
+    }
+
+    PyBaseExceptionGroupObject *self =
+        _PyBaseExceptionGroupObject_cast(BaseException_new(cls, args, kwds));
+    self->msg = Py_NewRef(msg);
+    self->excs = PySequence_Tuple(excs);
+    return (PyObject*)self;
+}
+
+static int
+BaseExceptionGroup_init(PyBaseExceptionGroupObject *self,
+    PyObject *args, PyObject *kwds)
+{
+    if (!_PyArg_NoKeywords(Py_TYPE(self)->tp_name, kwds)) {
+        return -1;
+    }
+    if (BaseException_init((PyBaseExceptionObject *)self, args, kwds) == -1) {
+        return -1;
+    }
+    return 0;
+}
+
+static int
+BaseExceptionGroup_clear(PyBaseExceptionGroupObject *self)
+{
+    Py_CLEAR(self->msg);
+    Py_CLEAR(self->excs);
+    return BaseException_clear((PyBaseExceptionObject *)self);
+}
+
+static void
+BaseExceptionGroup_dealloc(PyBaseExceptionGroupObject *self)
+{
+    _PyObject_GC_UNTRACK(self);
+    BaseExceptionGroup_clear(self);
+    Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static int
+BaseExceptionGroup_traverse(PyBaseExceptionGroupObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->msg);
+    Py_VISIT(self->excs);
+    return BaseException_traverse((PyBaseExceptionObject *)self, visit, arg);
+}
+
+static PyObject *
+BaseExceptionGroup_str(PyBaseExceptionGroupObject *self)
+{
+    if (self->msg && PyUnicode_CheckExact(self->msg)) {
+        Py_INCREF(self->msg);
+        return self->msg;
+    }
```

Comment:
Isn't it going to return the same value if you skip this? (Except if people set `self.args` I suppose -- in that case do we care?)

Reply:
See https://github.com/python/cpython/blob/6c6164ad8f99b3ba6d89a35cbcc03970320a38a5/Objects/exceptions.c#L114

If len(args) > 1 it returns str(args) so that would include the nested exceptions as well. I don't think we want that because the traceback prints them separately (and uses str()) for the exception_only part).

Reply:
I think you're right, message is set in new and is later read-only. I'll put asserts and remove the check.